File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.java
class Solution {
    static class Point {
        long distance;
        int[] point;

        public Point(long distance, int[] point) {
            this.distance = distance;
            this.point = point;
        }
    }

    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Point> pq = new PriorityQueue<>(Comparator.comparingLong(a -> a.distance));

        for (int[] point : points) {
            long distance = (long) point[0] * point[0] + (long) point[1] * point[1];
            pq.add(new Point(distance, point));
        }

        int[][] results = new int[k][];
        for (int i = 0; i < k; i++) {
            results[i] = pq.poll().point;
        }

        return results;
    }
}

File: 1014-k-closest-points-to-origin/1014-k-closest-points-to-origin.kt
class Solution {
    data class Point(var distance: Long, var point: IntArray)

    fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
        val pq: Queue<Point> = PriorityQueue(compareBy { a -> a.distance })

        for (point in points) {
            val distance = point[0].toLong() * point[0] + point[1].toLong() * point[1]
            pq.add(Point(distance, point))
        }

        val results: Array<IntArray> = Array(k) { IntArray(2) { 0 } }
        for (i in 0 until k) {
            results[i] = pq.poll().point
        }

        return results
    }
}